pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/SegmentSizeRule.java

```diff
@@ -82,20 +82,17 @@ public void run()
         && segmentSizeRuleParams.getNumRowsInActualSegment() == RecommenderConstants.SegmentSizeRule.NOT_PROVIDED) {
 
       // generate a segment
-      TableConfig tableConfig = createTableConfig(_input.getSchema());
-      int numRowsInGeneratedSegment = segmentSizeRuleParams.getNumRowsInGeneratedSegment();
-      File generatedSegmentDir =
-          new MemoryEstimator.SegmentGenerator(_input._schemaWithMetaData, _input._schema, tableConfig,
-              numRowsInGeneratedSegment, true).generate();
-      segmentSize = Math.round(FileUtils.sizeOfDirectory(generatedSegmentDir)
-          * RecommenderConstants.SegmentSizeRule.INDEX_OVERHEAD_RATIO_FOR_SEGMENT_SIZE);
-      numRows = numRowsInGeneratedSegment;
-
-      // cleanup
+      File workingDir = Files.createTempDir();
       try {
-        FileUtils.deleteDirectory(generatedSegmentDir);
-      } catch (IOException e) {
-        throw new RuntimeException("Cannot delete the generated segment directory", e);
+        TableConfig tableConfig = createTableConfig(_input.getSchema());
+        int numRowsInGeneratedSegment = segmentSizeRuleParams.getNumRowsInGeneratedSegment();
+        File generatedSegmentDir =
+            new MemoryEstimator.SegmentGenerator(_input._schemaWithMetaData, _input._schema, tableConfig,
+                numRowsInGeneratedSegment, true, workingDir).generate();
+        segmentSize = Math.round(FileUtils.sizeOfDirectory(generatedSegmentDir) * RecommenderConstants.SegmentSizeRule.INDEX_OVERHEAD_RATIO_FOR_SEGMENT_SIZE);
+        numRows = numRowsInGeneratedSegment;
+      } finally {
+        FileUtils.deleteQuietly(workingDir);
```

Comment:
We don't call cleanup from this rule but do so from RT provisioning rule. Let's have consistent semantics

Reply:
Please refer to previous comments.